File: py/focus/backup.py
"""Cópia do `focus.db` dentro do próprio acervo — espelha `src/backup.ts`.

Mesmo motivo do lado TypeScript: os arquivos capturados já estão a salvo em
`repository`, mas o catálogo (status de cada item, histórico de erro) só
existe em `BASE_DIR`, fora do disco de estudo. `VACUUM INTO` em vez de copiar
o arquivo cru porque o banco roda em WAL — copiar o `.db` bruto arriscaria uma
foto sem os dados ainda só no `.db-wal`.
"""
from __future__ import annotations

import sqlite3
import time

from . import config, db

COPIA_PATH = config.REPOSITORY / "focus.db"


def sincronizar_copia(conn: sqlite3.Connection) -> dict:
    inicio = time.monotonic()
    tmp = COPIA_PATH.with_name(COPIA_PATH.name + ".tmp")
    try:
        # VACUUM INTO recusa escrever num arquivo que já existe.
        tmp.unlink(missing_ok=True)
        alvo = str(tmp).replace("'", "''")
        conn.execute(f"VACUUM INTO '{alvo}'")
        tmp.replace(COPIA_PATH)
        ms = round((time.monotonic() - inicio) * 1000)
        return {"ok": True, "bytes": COPIA_PATH.stat().st_size, "ms": ms}
    except Exception as e:  # noqa: BLE001 — falha aqui não pode derrubar o worker
        ms = round((time.monotonic() - inicio) * 1000)
        db.log(conn, "error", "sync", f"cópia do banco falhou: {e}"[:300])
        return {"ok": False, "bytes": 0, "ms": ms, "erro": str(e)}

File: src/backup.ts
/**
 * Cópia do `focus.db` dentro do próprio acervo.
 *
 * `focus.db` mora em `BASE_DIR` — fora do disco de estudo, sem redundância
 * própria. Os arquivos capturados já estão a salvo em `repository`; falta o
 * catálogo que sabe o que cada um é, o que falta, e o que deu erro. Se
 * `BASE_DIR` se perder, este projeto se reconstrói do zero (`git clone`), mas
 * o catálogo — status de cada item, padrões do agente, histórico de erros —
 * não tem como ser refeito sem revisitar o site inteiro. Por isso a cópia.
 *
 * `VACUUM INTO` em vez de copiar o arquivo cru: o banco roda em WAL, então
 * parte dos dados mais recentes pode estar só no `.db-wal`, ainda não
 * migrada para o `.db` principal — copiar o arquivo bruto arriscaria uma
 * foto inconsistente. `VACUUM INTO` sempre lê a visão consolidada (WAL +
 * principal) e não precisa de exclusividade: outro processo pode estar
 * escrevendo ao mesmo tempo.
 *
 * A escrita vai para um `.tmp` e só então vira o nome final por `rename` —
 * uma cópia interrompida a meio nunca substitui a anterior, que ainda era boa.
 */
import { existsSync, renameSync, statSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import type { Database } from "bun:sqlite";

import { REPOSITORY } from "./config.ts";
import { log } from "./db.ts";

export const COPIA_PATH = join(REPOSITORY, "focus.db");

export interface Sincronizado {
  ok: boolean;
  bytes: number;
  ms: number;
  caminho: string;
  erro?: string;
}

export function sincronizarCopia(db: Database): Sincronizado {
  const inicio = Date.now();
  const tmp = `${COPIA_PATH}.tmp`;
  try {
    // VACUUM INTO recusa escrever num arquivo que já existe.
    if (existsSync(tmp)) unlinkSync(tmp);
    db.exec(`VACUUM INTO '${tmp.replace(/'/g, "''")}'`);
    renameSync(tmp, COPIA_PATH);
    return { ok: true, bytes: statSync(COPIA_PATH).size, ms: Date.now() - inicio, caminho: COPIA_PATH };
  } catch (e) {
    const erro = String(e).slice(0, 300);
    log(db, "error", "sync", `cópia do banco falhou: ${erro}`);
    return { ok: false, bytes: 0, ms: Date.now() - inicio, caminho: COPIA_PATH, erro };
  }
}
